File: backend/app/schemas/campaigns.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class CampaignCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    template_id: UUID
    smtp_account_id: UUID
    contact_list_id: UUID


class CampaignSummary(BaseModel):
    id: UUID
    name: str
    status: str
    total_recipients: int
    sent_count: int
    failed_count: int
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None


class CampaignDetail(CampaignSummary):
    template_id: UUID
    template_name: str
    smtp_account_id: UUID
    smtp_email: str
    list_id: UUID
    list_name: str
    pending_count: int


class CampaignCreateResponse(BaseModel):
    success: bool = True
    campaign_id: UUID


class CampaignListResponse(BaseModel):
    items: list[CampaignSummary]
    page: int
    limit: int
    total: int


class LaunchResponse(BaseModel):
    success: bool = True
    status: str


class CancelResponse(BaseModel):
    success: bool = True
    status: str


class DeleteResponse(BaseModel):
    success: bool = True

File: frontend/src/types/campaigns.ts
export type CampaignStatus =
  | "draft"
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export type CampaignSummary = {
  id: string;
  name: string;
  status: CampaignStatus;
  total_recipients: number;
  sent_count: number;
  failed_count: number;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
};

export type CampaignDetail = CampaignSummary & {
  subject: string | null;
  html_body: string | null;
  to_variable: string | null;
  selected_columns: string[] | null;
  smtp_account_id: string;
  smtp_email: string;
  list_id: string;
  list_name: string;
  pending_count: number;
};

export type CampaignListResponse = {
  items: CampaignSummary[];
  page: number;
  limit: number;
  total: number;
};

export type CampaignCreatePayload = {
  name: string;
  smtp_account_id: string;
  contact_list_id: string;
  subject: string;
  html_body: string;
  to_variable: string;
  selected_columns: string[];
};

export type CampaignPreviewPayload = {
  contact_list_id: string;
  subject: string;
  html_body: string;
  to_variable: string;
};

export type CampaignPreviewResult = {
  to: string;
  subject: string;
  html_body: string;
};
